File: src/main/java/net/rentalhost/plugins/php/hammer/inspections/codeStyle/NullableTypeRightmostInspection.java
package net.rentalhost.plugins.php.hammer.inspections.codeStyle;

import com.google.common.collect.Iterables;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.jetbrains.php.config.PhpLanguageLevel;
import com.jetbrains.php.lang.inspections.PhpInspection;
import com.jetbrains.php.lang.psi.elements.impl.PhpTypeDeclarationImpl;
import com.jetbrains.php.lang.psi.resolve.types.PhpType;

import java.util.stream.Collectors;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import net.rentalhost.plugins.services.LocalQuickFixService;
import net.rentalhost.plugins.services.ProblemsHolderService;
import net.rentalhost.plugins.services.TypeService;

public class NullableTypeRightmostInspection
    extends PhpInspection {
    @Override
    public @NotNull PsiElementVisitor buildVisitor(
        @NotNull final ProblemsHolder problemsHolder,
        final boolean isOnTheFly
    ) {
        return new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull final PsiElement element) {
                if (element instanceof PhpTypeDeclarationImpl) {
                    final var elementTypeText = element.getText();

                    if (!elementTypeText.startsWith("?")) {
                        final var elementTypes = TypeService.splitTypes(elementTypeText)
                                                            .map(s -> s.equals("null") ? PhpType._NULL : s)
                                                            .collect(Collectors.toList());

                        if (elementTypes.contains(PhpType._NULL) &&
                            !Iterables.getLast(elementTypes).equals(PhpType._NULL)) {
                            final var elementTypeReplacementSuggestion = TypeService.joinTypesStream(TypeService.exceptNull(elementTypeText)) + "|null";

                            ProblemsHolderService.registerProblem(
                                problemsHolder,
                                element,
                                String.format("Nullable type must be on rightmost side (\"%s\").", elementTypeReplacementSuggestion),
                                new LocalQuickFixService.SimpleTypeReplaceQuickFix(
                                    elementTypeReplacementSuggestion,
                                    "Move \"null\" type to rightmost side"
                                )
                            );
                        }
                    }
                }
            }
        };
    }

    @Override
    protected @Nullable PhpLanguageLevel getMinimumSupportedLanguageLevel() {
        return PhpLanguageLevel.PHP800;
    }
}

File: src/main/java/net/rentalhost/plugins/php/hammer/inspections/codeStyle/NullableTypeRightmostInspection.kt
package net.rentalhost.plugins.php.hammer.inspections.codeStyle

import com.google.common.collect.Iterables
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.jetbrains.php.config.PhpLanguageLevel
import com.jetbrains.php.lang.inspections.PhpInspection
import com.jetbrains.php.lang.psi.elements.impl.PhpTypeDeclarationImpl
import com.jetbrains.php.lang.psi.resolve.types.PhpType
import net.rentalhost.plugins.services.LocalQuickFixService.SimpleTypeReplaceQuickFix
import net.rentalhost.plugins.services.ProblemsHolderService
import net.rentalhost.plugins.services.TypeService
import kotlin.streams.toList

class NullableTypeRightmostInspection: PhpInspection() {
    override fun buildVisitor(problemsHolder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor = object: PsiElementVisitor() {
        override fun visitElement(element: PsiElement) {
            if (element is PhpTypeDeclarationImpl) {
                val elementTypeText = element.getText()

                if (elementTypeText.startsWith("?"))
                    return

                val elementTypes = TypeService.splitTypes(elementTypeText)
                    .map { s: String? -> if (s == "null") PhpType._NULL else s }
                    .toList()

                if (elementTypes.contains(PhpType._NULL) &&
                    Iterables.getLast(elementTypes) != PhpType._NULL) {
                    val elementTypeReplacementSuggestion = TypeService.joinTypesStream(TypeService.exceptNull(elementTypeText)) + "|null"

                    ProblemsHolderService.registerProblem(
                        problemsHolder,
                        element,
                        "Nullable type must be on rightmost side (\"$elementTypeReplacementSuggestion\").",
                        SimpleTypeReplaceQuickFix(
                            "Move \"null\" type to rightmost side",
                            elementTypeReplacementSuggestion
                        )
                    )
                }
            }
        }
    }

    override fun getMinimumSupportedLanguageLevel(): PhpLanguageLevel = PhpLanguageLevel.PHP800
}
